element_array_ephys/ephys_precluster.py

```diff
@@ -0,0 +1,892 @@
+        param_dict = {'precluster_method': precluster_method,
+                      'paramset_idx': paramset_idx,
+                      'paramset_desc': paramset_desc,
+                      'params': params,
+                      'param_set_hash':  dict_to_uuid(params)}
+        param_query = cls & {'param_set_hash': param_dict['param_set_hash']}
+
+        if param_query:  # If the specified param-set already exists
+            existing_paramset_idx = param_query.fetch1('paramset_idx')
+            if existing_paramset_idx == paramset_idx:  # If the existing set has the same paramset_idx: job done
+                return
+            else:  # If not same name: human error, trying to add the same paramset with different name
+                raise dj.DataJointError(
+                    'The specified param-set'
+                    ' already exists - paramset_idx: {}'.format(existing_paramset_idx))
+        else:
+            cls.insert1(param_dict)
+
+
+@schema
+class PreClusterParamList(dj.Manual):
+    definition = """
+    # Ordered list of paramset_idx that are to be run
+    precluster_param_list_id: smallint
+    """
+
+    class ParamOrder(dj.Part):
+        definition = """
+        -> master
+        order_id: smallint                  # Order of operations
+        ---
+        -> [nullable] PreClusterParamSet    # Nullable for when pre-clustering is not performed.
+        """
+
+
+@schema
+class PreClusterTask(dj.Manual):
+    definition = """
+    # Manual table for defining a clustering task ready to be run
+    -> EphysRecording
+    -> PreClusterParamList
+    ---
+    precluster_output_dir: varchar(255)  #  pre-clustering output directory relative to the root data directory
+    task_mode='none': enum('none','load', 'trigger') # 'none': no pre-clustering analysis
+                                                     # 'load': load analysis results
+                                                     # 'trigger': trigger computation
+    """
+
+
+@schema
+class PreCluster(dj.Imported):
+    """
+    A processing table to handle each PreClusterTask:
+    + If `task_mode == "none"`: no pre-clustering performed
+    + If `task_mode == "trigger"`: trigger pre-clustering analysis according to the      
+                                    PreClusterParamSet
+    + If `task_mode == "load"`: verify output
+    """
+    definition = """
+    -> PreClusterTask
+    ---
+    precluster_time: datetime  # time of generation of this set of pre-clustering results 
+    package_version='': varchar(16)
+    """
+
+    def make(self, key):
+        task_mode, output_dir = (PreClusterTask & key).fetch1('task_mode', 
+                                                              'precluster_output_dir')
+        precluster_output_dir = find_full_path(get_ephys_root_data_dir(), output_dir)
+
+        if task_mode == 'none':
+            creation_time = (EphysRecording & key).fetch1('recording_datetime')
```

Comment:
Throw an error if `task_mode=none` and there are entries in `PreClusterParamStep.Step`.

Reply:
Added `ValueError`.